cmd/oras/blob/push.go

```diff
@@ -0,0 +1,148 @@
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package blob
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/spf13/cobra"
+	"oras.land/oras/cmd/oras/internal/display"
+	"oras.land/oras/cmd/oras/internal/option"
+	"oras.land/oras/internal/file"
+)
+
+type pushBlobOptions struct {
+	option.Common
+	option.Descriptor
+	option.Pretty
+	option.Remote
+
+	fileRef   string
+	mediaType string
+	size      int64
+	targetRef string
+}
+
+func pushCmd() *cobra.Command {
+	var opts pushBlobOptions
+	cmd := &cobra.Command{
+		Use:   "push [flags] name[@digest] file",
+		Short: "[Preview] Push a blob to a remote registry",
+		Long: `[Preview] Push a blob to a remote registry
+
+** This command is in preview and under development. **
+
+Example - Push blob "hi.txt":
+  oras blob push localhost:5000/hello hi.txt
+
+Example - Push blob "hi.txt" with the specific digest:
+  oras blob push localhost:5000/hello@sha256:9a201d228ebd966211f7d1131be19f152be428bd373a92071c71d8deaf83b3e5 hi.txt
+
+Example - Push blob from stdin with blob size:
+  oras blob push --size 12 localhost:5000/hello -
+
+Example - Push blob "hi.txt" and output the descriptor:
+  oras blob push --descriptor localhost:5000/hello hi.txt
+
+Example - Push blob "hi.txt" with the specific returned media type in the descriptor:
+  oras blob push --media-type application/vnd.oci.image.config.v1+json --descriptor localhost:5000/hello hi.txt
+
+Example - Push blob "hi.txt" and output the prettified descriptor:
+  oras blob push --descriptor --pretty localhost:5000/hello hi.txt
+
+Example - Push blob without TLS:
+  oras blob push --insecure localhost:5000/hello hi.txt
+`,
+		Args: cobra.ExactArgs(2),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			opts.targetRef = args[0]
+			opts.fileRef = args[1]
+			if opts.fileRef == "-" && opts.PasswordFromStdin {
+				return errors.New("`-` read file from input and `--password-stdin` read password from input cannot be both used")
+			}
+			if opts.fileRef == "-" && opts.size == 0 {
+				return errors.New("`--size` must be provided if the blob is read from stdin")
+			}
+
+			return opts.ReadPassword()
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return pushBlob(opts)
+		},
+	}
+
+	cmd.Flags().Int64VarP(&opts.size, "size", "", 0, "provide the blob size")
+	cmd.Flags().StringVarP(&opts.mediaType, "media-type", "", ocispec.MediaTypeImageLayer, "specify the returned media type in the descriptor if `--descriptor` is used")
```

Comment:
@FeynmanZhou @yizha1 What's the default media type for blobs? Currently, it is `ocispec.MediaTypeImageLayer`, which is consistent with `oras push`. We need to review it later.